lib/SPIRV/SPIRVWriter.cpp

```diff
@@ -2072,9 +2072,11 @@ LLVMToSPIRVBase::transValueWithoutDecoration(Value *V, SPIRVBasicBlock *BB,
         IndexGroupArrayMap[ContainedIndexGroup].insert(AccessedArrayId);
       }
     }
-
-    SPIRVType *TranslatedTy = transPointerType(
-        GEP->getResultElementType(), GEP->getType()->getPointerAddressSpace());
+    SPIRVType *TranslatedTy =
+        GEP->getType()->isVectorTy()
+            ? transType(GEP->getType())
```

Comment:
This isn't right, though, in opaque-pointer mode, right? `GEP->getType()` would be `<4 x ptr>`, which would get translated as if it were `<4 x i8*>`. Instead, for vector types, you'd want to wrap the result of `transPointerType(...)` into a vector type--`transType(VectorType::get(TypedPointerType::get()))` is probably the cleanest way to get it. (Admittedly, vector-of-pointer types is likely to break elsewhere in opaque pointer mode anyways--the type scavenger for sure can't handle it right now).

Reply:
Thanks! Probably the latest code is the one we need.
There is some inefficiency though in opaque pointers mode: I see
```
    %1 = bitcast <4 x ptr addrspace(1)> %tmp2 to <4 x ptr addrspace(1)>
    %2 = bitcast <4 x ptr addrspace(1)> %1 to <4 x ptr addrspace(1)>
```
after reverse translation. The second bitcast relates with this patch, while the first one does not. May be it's worth to run DCE pass after the translation, since TBH I don't see a good way to eliminate these useless bitcasts.

Reply:
Fixed. The tests are now passing in debug as well. @jcranmer-intel please take another look when you have time.